util/merkle-tree/src/merkle_tree.rs

```diff
@@ -0,0 +1,354 @@
+use super::merge;
+use numext_fixed_hash::H256;
+use serde_derive::{Deserialize, Serialize};
+use std::collections::VecDeque;
+
+pub fn merkle_root(leaves: &[H256]) -> H256 {
+    let leaves_len = leaves.len();
+    // in case of empty slice, just return zero
+    if leaves_len == 0 {
+        return H256::zero();
+    }
+
+    let mut queue = VecDeque::with_capacity((leaves_len + 1) >> 1);
+
+    if leaves_len & 1 == 1 {
+        queue.push_back(leaves[0].clone());
+    }
+
+    let mut i = leaves_len;
+
+    while i > 1 {
+        i -= 2;
+        queue.push_back(merge(&leaves[i], &leaves[i + 1]));
+    }
+
+    while queue.len() > 1 {
+        let right = queue.pop_front().unwrap();
+        let left = queue.pop_front().unwrap();
+        queue.push_back(merge(&left, &right));
+    }
+
+    queue.pop_front().unwrap()
+}
+
+pub fn build_merkle_tree(leaves: Vec<H256>) -> Vec<H256> {
+    if leaves.is_empty() {
+        return vec![];
+    }
+
+    let mut nodes = vec![H256::zero(); leaves.len() - 1];
+    nodes.extend(leaves);
+    let mut i = nodes.len() - 1;
+
+    while i != 0 {
+        nodes[(i - 1) >> 1] = merge(&nodes[i - 1], &nodes[i]);
+        i -= 2;
+    }
+
+    nodes
+}
+
+pub fn update_root(tree: &mut Vec<H256>, mut index: usize) {
+    if index >= tree.len() {
+        return;
+    }
+
+    while index != 0 {
+        let sibling = ((index + 1) ^ 1) - 1;
+        let tmp = (index - 1) >> 1;
+        tree[tmp] = if sibling > index {
+            merge(&tree[index], &tree[sibling])
+        } else {
+            merge(&tree[sibling], &tree[index])
+        };
+        index = tmp;
+    }
+}
+
+#[derive(Eq, PartialEq, Debug, Serialize, Deserialize, Clone)]
+pub struct Proof {
```

Comment:
What's the difference between `Proof` and `CombinedProof`? I think that we can refactor the function to 
`build_proof(&self, indexes: &[usize]) -> Proof`, user can pass one item slice to build a single leaf proof: `tree.build_proof(&[1])`

Reply:
You can build a single node proof with CombinedProof, but using Proof, it will be less overhead.